Add jaccard calculation code (#521)
AlexsLemonade/scpca-nf: templates/qc_report/celltypes_supplemental_report.rmd

```

<!-- Define helper functions for calculating Jaccard similarity matrices --> 
```{r}
#' Function to calculate Jaccard similarity on two vectors
#'
#' @param vec1 First vector
#' @param vec2 Second vector
#'
#' @return Jaccard similarity between the vectors
jaccard <- function(vec1, vec2){
  length(intersect(vec1, vec2)) / length(union(vec1, vec2))
}

# Wrapper function to calculate jaccard similarity matrix for two categorical variables
#'
#' @param celltype_df The celltype_df data frame which must contain these columns:
#'   `colname1`, `colname2`, and `barcodes`
#' @param colname1 Column name, as a string, of first categorical variable of interest
#' @param colname2 Column name, as a string, of second categorical variable of interest
#'   
#' @return Jaccard similarity matrix for the two columns. `colname1` values will 
#'   be row names and `colname2` values will be column names in the final matrix
make_jaccard_matrix <- function(celltype_df, colname1, colname2){
  

>   # make lists of barcodes for each classifier's assignment, named by the classifier values

Reviewer: ```suggestion
  # make lists of barcodes for each category, named by the category
```
Since this function is generalized to categories, I think we can use that in the comment rather than classifier.

  id1_list <- split(celltype_df$barcodes, celltype_df[[colname1]])
  id2_list <- split(celltype_df$barcodes, celltype_df[[colname2]])
  
  # create the grid of comparisons
  cross_df <- tidyr::expand_grid(id1 = names(id1_list), id2 = names(id2_list))
  

>   # calculate scores using split lists & ids

Reviewer: ```suggestion
  # calculate a single Jaccard index for each combination using split lists & ids
```
Just a small clarification

#'   `colname1`, `colname2`, and `barcodes`
#' @param colname1 Column name, as a string, of first categorical variable of interest
#' @param colname2 Column name, as a string, of second categorical variable of interest
#'   
#' @return Jaccard similarity matrix for the two columns. `colname1` values will 
#'   be column names and `colname2` columns will be row names in the final matrix
make_jaccard_matrix <- function(celltype_df, colname1, colname2){
  
  # make lists of barcodes for each classifier's assignment, named by the classifier values
  id1_list <- split(celltype_df$barcodes, celltype_df[[colname1]])
  id2_list <- split(celltype_df$barcodes, celltype_df[[colname2]])
  
  # create the grid of comparisons
  cross_df <- tidyr::expand_grid(id1 = names(id1_list), id2 = names(id2_list))
  
  # calculate scores using split lists & ids
  jaccard_scores <- cross_df |> 
    purrr::pmap_dbl(\(id1, id2){
      jaccard(id1_list[[id1]], id2_list[[id2]])
    })
  
  # add scores to the comparison grid and convert to matrix
  jaccard_matrix <- cross_df |> 
    dplyr::mutate(jaccard = jaccard_scores) |>
    # convert to matrix
    tidyr::pivot_wider(

>       names_from = id1, 

Reviewer 2: Separately, I think we want to do the opposite matrix transformation here, so we get coordinates that are [id1, id2]

>       names_from = id1, 
>       # suppress warnings with dplyr::all_of (we're selecting with strings here)
>       values_from = dplyr::all_of("jaccard")

Reviewer 2: Selecting with just the strings ought to be fine here... 
```suggestion
      names_from = "id1", 
      values_from = "jaccard"
```

    ) |>
    tibble::column_to_rownames(var = "id2") |>
    as.matrix()
    
    return(jaccard_matrix)
}
```

```{r}
# Calculate all jaccard matrices of interest for input to heatmap

# baseline matrix
jaccard_mat <- matrix(
  nrow = 0,
  ncol = length(unique(celltype_df$cluster))
)

# vector to track splitting the heatmap
split_lengths <- c()  

if (has_submitter) {
  cluster_submitter_jaccard_matrix <- make_jaccard_matrix(celltype_df, "cluster", "submitter_celltype_annotation")

>   jaccard_mat <- rbind(jaccard_mat, cluster_submitter_jaccard_matrix)

Reviewer 2: I think we want to get the cluster names onto the rownames of this matrix somehow? I _think_ `split(celltype_df$cluster)` (which is effectively called in the make_jaccard_matrix) will always return the same order, which I would expect to be the same as `unique(celltype_df$cluster)`, but it might be worth using `join` to combine these matrices?

Author: They already are in the matrices from the jaccard calculation itself, with the `tibble::column_to_rownames(var = "id1")` code, unless I've misunderstood what you're asking here?

Reviewer 2: I think they would be lost in the `rbind`, but if we are not going to `rbind` and will use the ComplexHeatmap function instead, this will be irrelevant.
